prov/verbs_ud/src/verbs_ud.h

```diff
@@ -0,0 +1,738 @@
+
+static inline
+void ofi_verbs_ud_fork_init(void)
+{
+	int ret = FI_SUCCESS;
+
+	if (!gl_data.fork_unsafe) {
+		ret = ibv_fork_init();
+		if (ret) {
+			VERBS_UD_WARN(FI_LOG_CORE,
+				      "Enabling IB fork support failed: "
+				      "%s (%d)\n", strerror(ret), ret);
+		}
+		VERBS_UD_INFO(FI_LOG_CORE, "IB fork support "
+			      "is enabled\n");
+	} else {
+		VERBS_UD_INFO(FI_LOG_CORE, "IB fork support wasn't requested, "
+			      "not enabled it\n");
+	}
+}
+
+static inline
+int ofi_verbs_ud_have_device(void)
+{
+	struct ibv_device **devices;
+	struct ibv_context *verbs;
+	int found = 0, num_devices = 0, i;
+
+	devices = ibv_get_device_list(&num_devices);
+	if (!devices || !num_devices)
+		return found;
+
+	for (i = 0; i < num_devices; i++) {
+		verbs = ibv_open_device(devices[i]);
+		if (verbs) {
+			ibv_close_device(verbs);
+			found = 1;
+			break;
+		}
+	}
+
+	ibv_free_device_list(devices);
+	return found;
+}
+
+/* IPoIB devices */
+
+ofi_verbs_ud_infra_ipoib_t *ofi_verbs_ud_get_infra_ipoib(void);
+int ofi_verbs_ud_init_ipoib(void);
+int ofi_verbs_ud_add_addr_to_dev_list(struct dlist_entry *verbs_devs,
+				      struct sockaddr *addr);
+int ofi_verbs_ud_getinfo_ipoib(uint32_t version, const char *node,
+			       const char *service, uint64_t flags,
+			       struct fi_info *hints, struct fi_info **info);
+void ofi_verbs_ud_dev_cleanup_ipoib(void);
+void ofi_verbs_ud_fini_ipoib(void);
+
+/* ~IPoIB devices */
+
+/* All active hw devices */
+
+ofi_verbs_ud_infra_hw_t *ofi_verbs_ud_get_infra_hw(void);
+int ofi_verbs_ud_init_hw(void);
+int ofi_verbs_ud_getinfo_hw(uint32_t version, const char *node,
+			    const char *service, uint64_t flags,
+			    struct fi_info *hints, struct fi_info **info);
+void ofi_verbs_ud_dev_cleanup_hw(void);
+void ofi_verbs_ud_fini_hw(void);
+
+/* ~All actrive hw devices */
+
+/* ~Init functions */
+
+static inline int
+ofi_verbs_ud_get_port_info(struct ibv_context *context, int port,
+			   struct ibv_port_attr *attr)
+{
+	return ibv_query_port(context, port, attr);
+}
+
+static inline uint16_t
+ofi_verbs_ud_get_local_lid(struct ibv_context *context, int port)
+{
+	struct ibv_port_attr attr;
+
+	return ((ofi_verbs_ud_get_port_info(context, port, &attr)) ?
+		 0 : attr.lid);
+}
+
+#define OFI_INSTANTIATE_UTIL(prov_name, ofi_obj_name)			\
+static inline int ofi_##prov_name##_##ofi_obj_name##_instantiate(	\
+	struct ofi_##prov_name##_##ofi_obj_name **native_inst,		\
+	struct util_##ofi_obj_name **util_inst)				\
+{									\
+	*native_inst = (struct ofi_##prov_name##_##ofi_obj_name *)	\
+		calloc(1, sizeof(**native_inst));			\
+	if (!(*(native_inst)))						\
+		return -FI_ENOMEM;					\
+	if (util_inst)							\
+		*(util_inst) = &(*(native_inst))->util_##ofi_obj_name;	\
+	return FI_SUCCESS;						\
+}									\
+static inline void ofi_##prov_name##_##ofi_obj_name##_cleanup(		\
+	struct ofi_##prov_name##_##ofi_obj_name *native_inst)		\
+{									\
+	free(native_inst);						\
+}
```

Comment:
These macros don't look needed.  The second one in particular just obfuscates the code.